File: p3/TestDriver.java
/***********************************************
 * Author:     Malik Mohamedali
 * Assignment: Program 3
 * Class:       CSI 4337
 ************************************************/

import java.util.concurrent.*;

/**
 * "randomly" tests activity of consumer and producer threads with various habits
 */
public class TestDriver {
    /**
     * mandate thready activity and exit after X seconds
     * @param args none
     * @throws InterruptedException if thread in threadpool cannot wait
     */
    public static void main(String[] args) throws InterruptedException {
        final Buffet buffet = new BuffetSemaphore();
        // final Buffet buffet = new BuffetMonitor();
        //final Buffet buffet = new BuffetReentrantLock();

        int numProd = 20; // assign prods and consumers
        int numCons = 40;
        ThreadPoolExecutor prodThreads = (ThreadPoolExecutor) Executors.newFixedThreadPool(numProd);
        ThreadPoolExecutor consThreads = (ThreadPoolExecutor) Executors.newFixedThreadPool(numCons);

        long start = System.currentTimeMillis();
        long end = start + (10 * 1000); /// end after 10s of execution
        prodThreads.setKeepAliveTime(1, TimeUnit.SECONDS);
        consThreads.setKeepAliveTime(1, TimeUnit.SECONDS);
        prodThreads.allowCoreThreadTimeOut(true);
        consThreads.allowCoreThreadTimeOut(true);

        for (int i = 0; i < numProd; i++) { // initialize all producer threads
            // execute random producer thread
            SliceType type = SliceType.values()[ThreadLocalRandom.current().nextInt(0,SliceType.values().length)];
            int numSlices = ThreadLocalRandom.current().nextInt(1,2); // try and add 1 slices
            prodThreads.execute(new AddSlices(buffet, type, numSlices, end - 1000)); /// FIXME implement runnable to mandate call to buffet methods
        }
        for (int i = 0; i < numCons; i++) { // initialize all consumer threads
            // execute random consumer thread
            SliceType type = SliceType.values()[ThreadLocalRandom.current().nextInt(0,SliceType.values().length)];
            int numSlices = ThreadLocalRandom.current().nextInt(1,6); // try and take between 1 and 5 slices wanted
            consThreads.execute(new TakeSlices(buffet, type, numSlices, end - 1000)); /// FIXME implement runnable to mandate call to buffet methods
        }
        while (System.currentTimeMillis() < end); // allow prod/cons threads to execute for 30s before closing
        buffet.close();
        // Thread.sleep(2000); // delay main so child threads finish
    }
}

File: p3/testdriver.cpp
#include <stdio.h>
#include <pthread.h>
#include "ThreadProcs.cpp"

using namespace std;
using chrono::seconds;

Buffet BUFFET(20);

int main() {
    const int numThreads = 4;
    const int numProd = numThreads, numCons = numThreads; // assign prods and consumers
    srand((unsigned int)time(0));

    pthread_t prodIds[numProd];
    pthread_t consIds[numCons];
    chrono::steady_clock::time_point start = chrono::steady_clock::now();

    struct threadArgs *Pargs = new struct threadArgs[numThreads];
    struct threadArgs *Cargs = new struct threadArgs[numThreads];
    for (int i = 0; i < numThreads; i++) { // create child threads with arguments passed through void* struct
        Pargs[i].type = SliceType(i); // 1 producer of each type
        Pargs[i].numSlices = (rand() % 5) + 1; // add between 1 and 4 slices

        if (pthread_create(prodIds + i, NULL, prodProcess, Pargs + i) != 0) {
            fprintf(stderr, "Bad thread\n");
            exit(5);
        }
        Cargs[i].type = SliceType(rand() % 4); // random consumer type
        Cargs[i].numSlices = (rand() % 5) + 1; // add between 1 and 4 slices & 1 consumer of each type

        if (pthread_create(consIds + i, NULL, consProcess, Cargs + i) != 0) {
            fprintf(stderr, "Bad thread\n");
            exit(5);
        }
    }

    while (chrono::steady_clock::now() - start < chrono::seconds(20)) {} // run for 15s
    BUFFET.close(); //close buffet
    for (int i = 0; i < numThreads; i++) { // wait for all running children
        pthread_join(prodIds[i], NULL);
        pthread_join(consIds[i], NULL);
    }
    delete [] Cargs;
    delete [] Pargs;
    return 0;
}
